src/ipc/ipc-zephyr.c

```diff
@@ -59,33 +63,105 @@ LOG_MODULE_DECLARE(ipc, CONFIG_SOF_LOG_LEVEL);
  */
 static uint32_t g_last_data, g_last_ext_data;
 
+struct k_sem *wait_ack_sem;
+
 /**
- * @brief cAVS IPC Message Handler Callback function.
+ * @brief cAVS IPC Message Received Callback function.
  *
- * See @ref (*intel_adsp_ipc_handler_t) for function signature description.
- * @return false so BUSY on the other side will not be cleared immediately but
+ * @return -1 so BUSY on the other side will not be cleared immediately but
  * will remain set until message would have been processed by scheduled task, i.e.
  * until ipc_platform_complete_cmd() call.
  */
-static bool message_handler(const struct device *dev, void *arg, uint32_t data, uint32_t ext_data)
+static void ipc_receive_cb(const void *data, size_t len, void *priv)
 {
-	struct ipc *ipc = (struct ipc *)arg;
+	struct intel_adsp_ipc_ept_priv_data *priv_data =
+		(struct intel_adsp_ipc_ept_priv_data *)priv;
 
-	k_spinlock_key_t key;
+	if (len == INTEL_ADSP_IPC_CB_MSG) {
```

Comment:
`size_t len` is an `enum`??... I see that it comes from https://github.com/zephyrproject-rtos/zephyr/pull/96233/commits/24020058e6afd2ce12fa549122bc45dc71fb1206 and I don't find that a good idea... Can we change that to a simple (possibly `unsigned`) `int`?

Reply:
The `size_t` is from the API which I have no control of.

As for the `enum`, in a sense it is doing some "grouping" of values together. So, this "set" is for sending message, another set is for callback, etc. I can change it there if you insist. This can be done when this is merged and we go back to remove the old driver.

Reply:
@dcpleung hm, but that's an Intel-ADSP private API, so its only user is almost certainly SOF.. Do the same strict policies apply to it making changing it too difficult? At least we should change the name - it isn't really `len` any more? And changing `size_t` to a compatible explicit integer type should be doable too?

Reply:
I see. I have renamed the variable to `cb_type`. But I cannot change the data type since it's in the callback prototype.